oss-internship-2020/openjpeg/examples/decompress_example.cc

```diff
@@ -0,0 +1,188 @@
+  if (!stream.ok()) {
+    LOG(FATAL) << "opj_stream initialization failed: " << stream.status();
+    return EXIT_FAILURE;
+  }
+  sapi::v::RemotePtr stream_pointer(stream.value());
+
+  sapi::StatusOr<opj_codec_t*> codec = api.opj_create_decompress(OPJ_CODEC_JP2);
+  if (!codec.ok()) {
+    LOG(FATAL) << "opj_codec initialization failed: " << codec.status();
+    return EXIT_FAILURE;
+  }
+  sapi::v::RemotePtr codec_pointer(codec.value());
+
+  sapi::v::Struct<opj_dparameters_t> parameters;
+  status = api.opj_set_default_decoder_parameters(parameters.PtrBoth());
+  if (!status.ok()) {
+    LOG(FATAL) << "parameters initialization failed: " << status;
+    return EXIT_FAILURE;
+  }
+
+  sapi::StatusOr<OPJ_BOOL> bool_status =
+      api.opj_setup_decoder(&codec_pointer, parameters.PtrBefore());
+  if (!bool_status.ok() || !bool_status.value()) {
+    LOG(FATAL) << "decoder setup failed";
+    return EXIT_FAILURE;
+  }
+
+  // start reading image from the input file
+  sapi::v::GenericPtr image_pointer;
+  bool_status = api.opj_read_header(&stream_pointer, &codec_pointer,
+                                    image_pointer.PtrAfter());
+  if (!bool_status.ok() || !bool_status.value()) {
+    LOG(FATAL) << "reading image header failed";
+    return EXIT_FAILURE;
+  }
+
+  sapi::v::Struct<opj_image_t> image;
+  image.SetRemote((void*)image_pointer.GetValue());
+  if (!sandbox.TransferFromSandboxee(&image).ok()) {
+    LOG(FATAL) << "transfer from sandboxee failed";
+    return EXIT_FAILURE;
+  }
+
+  sapi::v::RemotePtr image_remote_pointer(image.GetRemote());
+  bool_status =
+      api.opj_decode(&codec_pointer, &stream_pointer, &image_remote_pointer);
+  if (!bool_status.ok() || !bool_status.value()) {
+    LOG(FATAL) << "decoding failed";
+    return EXIT_FAILURE;
+  }
+
+  bool_status = api.opj_end_decompress(&codec_pointer, &stream_pointer);
+  if (!bool_status.ok() || !bool_status.value()) {
+    LOG(FATAL) << "ending decompress failed";
+    return EXIT_FAILURE;
+  }
+
+  int components = image.data().numcomps;
+
+  // transfer the read data to the main process
+  sapi::v::Array<opj_image_comp_t> image_components(components);
+  image_components.SetRemote(image.data().comps);
+  if (!sandbox.TransferFromSandboxee(&image_components).ok()) {
+    LOG(FATAL) << "transfer from sandboxee failed";
+    return EXIT_FAILURE;
+  }
+
+  image.mutable_data()->comps = (opj_image_comp_t*)image_components.GetLocal();
+
+  int width = (int)image.data().comps[0].w;
+  int height = (int)image.data().comps[0].h;
+
+  std::vector<std::vector<OPJ_INT32>> data;
+  sapi::v::Array<OPJ_INT32> image_components_data(width * height);
+
+  for (int i = 0; i < components; ++i) {
+    image_components_data.SetRemote(image.data().comps[i].data);
+    if (!sandbox.TransferFromSandboxee(&image_components_data).ok()) {
+      LOG(FATAL) << "transfer from sandboxee failed";
+      return EXIT_FAILURE;
+    }
+    std::vector<OPJ_INT32> component_data(
```

Comment:
Use `emplace_back` or `std::move()` the temporary into place or use unnamed temporary inside `push_back`.